import "CommonTopology.proto";

package SwitchRegistryModule_Proto;

option java_package = "zsdn.protocol";
option java_outer_classname = "SwitchRegistryModuleProtocol";


// Main message type for messages FROM SwitchRegistryModule

message From {


    oneof FromMsg {
        SwitchEvent switch_event = 1;
    }

    message SwitchEvent{
        oneof SwitchEventType {
            common.topology.Switch switch_added = 1;
            common.topology.Switch switch_removed = 2;
            SwitchChanged switch_changed = 3;
        }
        message SwitchChanged {
            required common.topology.Switch switch_before = 1;
            required common.topology.Switch switch_now = 2;
        }
    }


}

// All Requests the SwitchRegistryModule will handle

message Request  {

    oneof RequestMsg {
        GetAllSwitchesRequest get_all_switches_request = 1;
        GetSwitchByIdRequest get_switch_by_id_request = 2;
    }



    // Request submessage definitions

    message GetAllSwitchesRequest {}

    message GetSwitchByIdRequest {
        required uint64 switch_dpid = 1;
    }
}





// All replies the SwitchRegistryModule can send

message Reply {

    oneof ReplyMsg {
        GetAllSwitchesReply get_all_switches_reply = 1;
        GetSwitchByIdReply get_switch_by_id_reply = 2;
    }



    // Reply submessage definitions

    message GetAllSwitchesReply {
        repeated common.topology.Switch switches = 1;
    }

    message GetSwitchByIdReply {
        optional common.topology.Switch switch = 1;
    }
}
